### PR TITLE
Rewrite ExprResolver with pattern matching

### PR DESCRIPTION
**Pull request by imkiva**
Rewrite ExprResolver with pattern matching

### CODE REVIEW

File: base/src/main/java/org/aya/resolve/visitor/ExprResolver.java [hunk 1]
```diff
@@ -29,112 +30,167 @@
...
-    return expr;
-  }
-
-  @Override public @NotNull Expr visitMetaPat(Expr.@NotNull MetaPat metaPat, Context p) {
-    return metaPat;
-  }
+  @Override @NotNull ExprView view,
+  @NotNull State state,
+  @NotNull Context context
+) implements ExprOps {
```
Comment (ice1000) [inline base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:41]: Does it have to be an implementation of `ExprOps`?

**Review by ice1000 — COMMENTED** [review]
To kiva chan 's question

File: base/src/main/java/org/aya/resolve/visitor/ExprResolver.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:41]: I think probably no. And I even think that it's not necessary to rewrite the expr resolver with ExprView (at least right now):

Comment (imkiva) [reply base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:41]: But [this commit](https://github.com/aya-prover/aya-dev/pull/374/commits/bdbce8ceb33f8035747d99d39958ca0b437cfba9) is still useful and I would like to give it a cherry-pick 🍒 .

Comment (ice1000) [reply base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:41]: > But [this commit](https://github.com/aya-prover/aya-dev/pull/374/commits/bdbce8ceb33f8035747d99d39958ca0b437cfba9) is still useful and I would like to give it a cherry-pick 🍒 .

Of course, I would like to merge the first commit and the part that you mark yourself as an author

Comment (ice1000) [reply base/src/main/java/org/aya/resolve/visitor/ExprResolver.java:41]: bors d=imkiva

**Review by ice1000 — APPROVED** [review]
Rest lgtm

**Comment by imkiva** [comment]
bors merge

**Pull request merged**